code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/LoadBalancerServiceCreate.java

```diff
@@ -0,0 +1,183 @@
+            data.put(LoadBalancerConstants.FIELD_LB_NETWORK_ID, sdService.getServiceNetworkId(service));
+            data.put(
+                    LoadBalancerConstants.FIELD_LB_INSTANCE_IMAGE_UUID,
+                    launchConfigData.get(InstanceConstants.FIELD_IMAGE_UUID));
+            data.put(
+                    LoadBalancerConstants.FIELD_LB_INSTANCE_URI_PREDICATE,
+                    DataAccessor.fields(service).withKey(LoadBalancerConstants.FIELD_LB_INSTANCE_URI_PREDICATE)
+                            .withDefault("delegate:///").as(
+                                    String.class));
+            data.put("accountId", service.getAccountId());
+            lb = objectManager.create(LoadBalancer.class, data);
+        }
+
+        objectProcessManager.executeProcess(LoadBalancerConstants.PROCESS_LB_CREATE, lb, null);
+    }
+
+    private LoadBalancerConfig createDefaultLoadBalancerConfig(String defaultName, long accountId,
+            Map<String, Object> lbConfigData) {
+        String name = lbConfigData.get("name") == null ? defaultName : lbConfigData.get("name")
+                .toString();
+        LoadBalancerConfig lbConfig = objectManager.findOne(LoadBalancerConfig.class,
+                LOAD_BALANCER_CONFIG.NAME,
+                name,
+                LOAD_BALANCER_CONFIG.REMOVED, null,
+                LOAD_BALANCER_CONFIG.ACCOUNT_ID, accountId);
+
+        if (lbConfig == null) {
+            lbConfigData.put("accountId", accountId);
+            lbConfigData.put("name", name);
+            lbConfig = resourceDao.createAndSchedule(LoadBalancerConfig.class, lbConfigData);
+        }
+        return lbConfig;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void createListeners(Service service, LoadBalancerConfig lbConfig, Map<String, Object> launchConfigData) {
+        // 1. create listeners
+        List<String> portDefs = (List<String>) launchConfigData.get(InstanceConstants.FIELD_PORTS);
+        if (portDefs == null || portDefs.isEmpty()) {
+            return;
+        }
+
+        Map<Integer, LoadBalancerListener> listeners = new HashMap<>();
+
+        for (String port : portDefs) {
+            PortSpec spec = new PortSpec(port);
+            if (!port.contains("tcp")) {
+                // default to http unless defined otherwise in the compose file
+                spec.setProtocol("http");
+            }
+
+            if (listeners.containsKey(spec.getPrivatePort())) {
+                continue;
+            }
+            
+            Integer publicPort = spec.getPublicPort();
+            int privatePort = spec.getPrivatePort();
+            if (publicPort == null) {
+                publicPort = privatePort;
+            }
+
+            LoadBalancerListener listenerObj = objectManager.findOne(LoadBalancerListener.class,
+                    LOAD_BALANCER_LISTENER.NAME, sdService.getLoadBalancerName(service) + "_" + publicPort,
```

Comment:
Craig, initially I've made them not editable. But then Vince and Darren suggested I allow to edit them. So you are right, for lb service I should no longer rely on the name of the listener/config to be consistent over time. I will add serviceId to indicate they are the part of the service (like I already do for the load balancer itself)